CI: Run coccicheck BPF target with GitHub Actions

pchaigno opened the pull request:
Runs `coccicheck` as a GitHub Action on all pull request synchronize and open events, with Coccinelle packaged as a Docker image based on Alpine.

pchaigno committed:
coccinelle: Run coccicheck with GitHub Actions

Runs coccicheck as a GitHub Action on all pull request synchronize and
open events, with Coccinelle packaged as a Docker image based on Alpine.

Signed-off-by: Paul Chaignon <paul@cilium.io>

qmonnet commented on .github/workflows/coccicheck.yaml:
```diff
@@ -0,0 +1,12 @@
+jobs:
+    coccicheck:
+        name: coccicheck
+        runs-on: ubuntu-18.04
```
Nit: Looking at `go-mod.yaml`, it looks like we could also have `ubuntu-latest` instead of 18.04. Probably doesn't change much though.

qmonnet commented on contrib/coccinelle/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+LABEL maintainer="maintainer@cilium.io"
+
+RUN apk add curl autoconf automake gcc libc-dev ocaml ocaml-dev ocaml-ocamldoc ocaml-findlib make python bash && \
```
I looked at [this Dockerfile](https://hub.docker.com/r/neszt/cppcheck-docker/dockerfile) and the maintainer seems to use a `-t name` option to name the list of packages, and delete them cleanly at the end of the sequence. Not sure if applies here given that some packages are not deleted, but might be worth checking how that option works.

Also in the same Dockerfile, the compiled binary is stripped and `/usr/src` is removed, to minimise the size of the Docker image. I have little knowledge of best practices for Docker, but shrinking the image size does sound like a good thing, so maybe we could make the same here?

pchaigno replied:
> the maintainer seems to use a -t name option to name the list of packages, and delete them cleanly at the end of the sequence.

Oh, that's neat! I'll use it.

> the compiled binary is stripped

That saves 3.8/64.9MB. Done.

> `/usr/src` is removed.

Doesn't seem to make a difference in our case.

qmonnet commented on .github/workflows/coccicheck.yaml:
```diff
@@ -0,0 +1,12 @@
+name: BPF
```
```suggestion
name: BPF checks
```

qmonnet reviewed (commented):
Looks good!

I started to look at running checkers for BPF in a container (thinking of coccichecks, but also checkpatsh.pl, sparse, clang, cppcheck). I thought we could run all of them in the same container, to avoid pulling too many things.

I thought we would add a GitHub and a Docker image with all checkers at once, but it's probably not a bad thing to start with just one. The only nit is that we may have to rename the files (`.github/workflows/coccicheck.yaml`),  job and Docker image if we add new tools. But that should not be an issue, what do you think?

tklauser commented on .github/workflows/coccicheck.yaml:
```diff
@@ -0,0 +1,12 @@
+        name: coccicheck
+        runs-on: ubuntu-18.04
+        steps:
+            - uses: actions/checkout@v1
```
Similar to @qmonnet's comment above, we could use `actions/checkout@v2` like in `go-mod.yaml`. Also see https://github.com/actions/checkout#whats-new

```suggestion
            - uses: actions/checkout@v2
```

pchaigno replied:
Thanks for including the link!

tklauser commented on contrib/coccinelle/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+LABEL maintainer="maintainer@cilium.io"
+
+RUN apk add curl autoconf automake gcc libc-dev ocaml ocaml-dev ocaml-ocamldoc ocaml-findlib make python bash && \
+    curl -sS -L https://github.com/coccinelle/coccinelle/archive/1.0.8.tar.gz -o coccinelle-1.0.8.tar.gz && \
```
Might be worth to declare an `ENV COCCINELL_VERSION 1.0.8` at the top of the `Dockerfile` and use it here and below, so future version bumps are easier.

tklauser reviewed (commented):
Nice! Two rather small comments inline, feel free to ignore.

pchaigno committed:
bpf: Mark a few parameters as const

Changes in this commit were generated with the following command,
provided by the coccicheck CI:

  spatch --sp-file contrib/coccinelle/const.cocci --include-headers \
    --very-quiet --in-place bpf/

Signed-off-by: Paul Chaignon <paul@cilium.io>

tklauser reviewed (approved):
Nice :tada:

borkmann reviewed (approved):
Nice, thanks!
